### ISSUE TITLE
CultureInsensitiveTypeAttribute is no solution for libraries

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#901]
If you have a library/assembly for which you have no control on the clients, the use of **CultureInsensitiveTypeAttribute** is no solution for suppressing **MA0011** / **MA0076** for a number of formats.

The **CultureInsensitiveTypeAttribute** annotations that are defined in the library are not "baked" into that assembly (due to the **MEZIANTOU_ANALYZER_ANNOTATIONS** conditional symbol). This means that they do not exist for **Menziantoo.Analyzer** when analyzing a project that uses the library on which they are defined.

Unless I'm missing something, we'd have to inform each consumer of our libraries (that uses **Menziantoo.Analyzer**) to define the invariant formats using assembly-level **CultureInsensitiveTypeAttribute** annotations in their project(s). That doesn't seem like a good or easy to maintain solution.

It works perfectly fine if you want to suppress these diagnostics within the library in which the type is defined though.

Should we go back to the "suppressor" option? For that option, we'd almost have to implement the same logic  that **Menziantoo.Analyzer** uses to detect whether a type or format is invariant. If not, we risk suppressing too much.

This is mostly problematic in case of **MA0011**:

```c#
using System;
using System.Text;

internal sealed class Program
{
    private static void Main(string[] args)
    {
        Foo foo = new Foo();

        var sb = new StringBuilder();
        sb.Append($"{5:x}").Append($"{foo:o} {DateTime.Now:D}").Append("!!"); // MA0011 should be reported because the "D" format of DateTime is not culture-invariant
        sb.Append($"{5:x}").Append($"{foo:o} {DateTime.Now:o}").Append("!!"); // MA0011 should be suppressed
    }
}

public sealed class Foo : IFormattable
{
    public string ToString(string? format)
    {
        return string.Empty;
    }

    public string ToString(string? format, IFormatProvider? formatProvider)
    {
        return string.Empty;
    }

    public override string ToString()
    {
        return string.Empty;
    }
}
```

If we assume that the "o" format for **Foo** is culture-invariant, then we'd only want **MA0011**  to be reported for `DateTime.Now:D`.

The diagnostic that **MA0011** reports is for the whole line up to where the issue is situated. For the example above, `sb.Append($"{5:x}").Append($"{foo:o} {DateTime.Now:D}")` is the location for which **MA0011** is reported.

This means that for `sb.Append($"{5:x}").Append($"{foo:o} {DateTime.Now:o}")` we'd have to know that the "o" format for **DateTime** and the "x" format for **System.Int32** are culture-invariant and effectively suppress **MA0011**.

For `sb.Append($"{5:x}").Append($"{foo:o} {DateTime.Now:D}")` we'd have to know that the "D" format for **DateTime** is not culture-variant and **NOT** suppress **MA0011**.

Not that important, but why isn't **MA0011** only reported for `Append($"{foo:o} {DateTime.Now:D}")` in the example above?

For us, it would even be better if **MA0011** were to be reported twice: once with `{foo:o}` as location and once with `{DateTime.Now:D}` as location. This would allow us to easily suppress the diagnostic for `{foo:o}`.

I doubt this is acceptable for you though, and I perfectly understand.

**Comment by drieseng** [comment]
@meziantou, I was able to resolve it by defining the **CultureInsensitiveTypeAttribute** as internal class in the library. Is that the best approach?

**Comment by meziantou** [comment]
- A library publisher can use add annotations and add the `MEZIANTOU_ANALYZER_ANNOTATIONS` to their csproj to preserve the annotation after the compilation.
- As a consumer, you can use `[assembly: CultureInsensitiveType(typeof(Foo), "format")]` to add annotation for a type that is not part of the code
- If this is a builtin type, let's update the analyzer to handle it (create issues)

> I was able to resolve it by defining the CultureInsensitiveTypeAttribute as internal class in the library. Is that the best approach?

It should work if you don't want to have a dependency on the annotation package.

**Comment by drieseng** [comment]
> It should work if you don't want to have a dependency on the annotation package.

OK, I'll apply that solution.

**Comment by drieseng** [comment]
I was too quick in closing this issue.
I cannot get it to work when **CultureInsensitiveTypeAttribute** is defined in the class library project.
You can reproduce it by cloning and build this repo:
https://github.com/drieseng/meziantoo_invariant.git

**Comment by drieseng** [comment]
It works fine if I make the **CultureInsensitiveTypeAttribute** public in the class library, but - if possible - I'd like to avoid that.

**Comment by meziantou** [comment]
I will to check what information are exposed in the analyzer. For instance, I know private symbols are not loaded to reduce resources as they are not needed. I wonder if this is the same here.

**Comment by meziantou** [comment]
It seems now ok with the latest version of the package. Can you validate it in your project and confirm if the fix is good?

**Comment by drieseng** [comment]
It's working great now, thanks!!!

I'll do some tests next week to see how these recent changes affect build time.

Would you consider introducing a **CultureInsensitiveFormatAttribute** and only use **CultureInsensitiveTypeAttribute** if you actually want to mark a given type as being culture-invariant? Meaning all format strings for the annotated type are to be considered culture invariant.

I'd be happy to submit a PR for this, but no problem if you prefer to have a single custom attribute of course.

**Comment by meziantou** [comment]
Hi,
Can you clarify why you need another attribute with a simple use-case?

**Comment by drieseng** [comment]
**CultureInsensitive(Type)Attribute** would be used to indicate that all formats (including the "default") of a given type are culture-invariant where as **CultureInsensitiveFormatAttribute** would be used to specify that specific formats are culture-invariant.

All formats of type **Foo** are invariant:
```c#
[CultureInsensitive(Type)]
public class Foo
{
    ...
}
```

Only the "o" and "x" format of type **Foo** are invariant:
```c#
[CultureInsensitiveFormat("o")]
[CultureInsensitiveFormat("x")]
public class Foo
{
    ...
}
```

I'm of course ok to leave it as is, if you don't think it's any clearer.

**Comment by meziantou** [comment]
So, if I understand correctly, you suggest splitting the existing attribute `CultureInsensitiveType` into two attributes.

````c#
// currently supported

// all formats are culture-insentive
[CultureInsensitiveType] class Foo;
// Specific formats are culture-insentive
[CultureInsensitiveType(format: "o")] class Foo;

// Suggested

// all formats are culture-insentive
[CultureInsensitive] class Foo;
// Specific formats are culture-insentive
[CultureInsensitiveFormat("o")] class Foo;
````

It would need support for 2 new attributes in addition to the current attribute to avoid the breaking change. Are there cases with the current solution (using a single attribute) that don't work well? Or is it just about renaming for better readability?

Note that adding new attributes in addition to the existing one may slow down the analyzer. Not by a lot, but this is something to consider.

**Comment by drieseng** [comment]
@meziantou, I'm ok to leave it as is. Thanks for considering it though.

I still need to evaluate the impact of recent changes on build time, but - if necessary - I'll open a separate issue for this.

Again a big thanks!

**Comment by drieseng** [comment]
@meziantou, the impact on build time is very limited:

**Before** (v2.0.215):

<img width="827" height="276" alt="Image" src="https://github.com/user-attachments/assets/5b2d4e81-2014-4b74-ad87-ee5db5653382" />

**After** (v2.0.250):

<img width="824" height="288" alt="Image" src="https://github.com/user-attachments/assets/10f21ecf-234e-4157-a216-da7ca0e1e723" />
